pkgs/development/libraries/wolfssl/default.nix

```diff
@@ -6,73 +6,119 @@
 , util-linux
 , openssl
 }:
-
-stdenv.mkDerivation rec {
-  pname = "wolfssl";
-  version = "5.6.3";
-
-  src = fetchFromGitHub {
+let
+  defaultVersion = "5.6.3";
+  defaultSrc = fetchFromGitHub {
     owner = "wolfSSL";
     repo = "wolfssl";
-    rev = "refs/tags/v${version}-stable";
+    rev = "refs/tags/v${defaultVersion}-stable";
     hash = "sha256-UN4zs+Rxh/bsLD1BQA+f1YN/UOJ6OB2HduhoetEp10Y=";
   };
 
-  postPatch = ''
-    patchShebangs ./scripts
-    # ocsp tests require network access
-    sed -i -e '/ocsp\.test/d' -e '/ocsp-stapling\.test/d' scripts/include.am
-    # ensure test detects musl-based systems too
-    substituteInPlace scripts/ocsp-stapling2.test \
-      --replace '"linux-gnu"' '"linux-"'
-  '';
+  buildWolfSSL =
+    { version ? defaultVersion
+    , src ? defaultSrc
+    , variant
+    , defaultFlags ? true
+    , enableOddCrypto ? false
+    , extraOptions ? [ ]
+    , extraFlags ? [ ]
+    , enableLto ? true
+    , extraCFLAGS ? ""
+    , extraLDFLAGS ? ""
```

Comment:
Ok, let me put it a bit more firmly - this is not how to do things like this. The override mechanisms should be perfectly capable of producing the effects of all of this, without fugly flags like `defaultFlags`, `extraOptions`, `extraLDFlags` that users have to try and understand the behaviour of.

If this were the way to achieve things like this, we'd have thousands of `build*` functions kicking around `pkgs`.

For example, the effect of `extraLDFlags` can be achieved by something like:

```
wolfssl.overrideAttrs (prevAttrs: {
  env = prevAttrs.env // {
    NIX_LDFLAGS_COMPILE = prevAttrs.env.NIX_LDFLAGS_COMPILE or [] ++ ["-foo"];
  };
})
```

which also gives the overrider the ability to choose whether their flags come before or after the existing flags.. or replace the existing flags entirely (which also does away with the need for weird arguments like `defaultFlags`)

Reply:
Sorry, I don't mean to ignore your feedback or waste your time.

Do you want me to remove just the extra* flags or the entire helper?

I do want the helper, as the 
`--enable-all` build will not work for my intended consumer (HAProxy), so I'd need to either duplicate all the flags or filter out the `--enable-all` flag and add `--enable-haproxy`.

Reply:
Indeed please ignore my unnecessary grumpiness.

With a combination of `.override` and `.overrideAttrs` you shouldn't need the helper at all. Something like `variant` could become a file-level argument with a default, allowing it to be overridden using `.override`. Perhaps `enableOddCrypto` could become a file-level argument too, but it could do with a better name because it's not obvious what it means ("odd" and "crypto" in the same sentence is a bit scary to some people).

`extraOptions` and `extraFlags` are achievable using `.overrideAttrs`. `defaultFlags` is unnecessary because they can all be removed through use of `.overrideAttrs`. `extraCFlags` and `extraLDFlags` are achievable using `.overrideAttrs`.

`enableLTO` is arguably worth making a file-level argument, overridable using `.override` because it translates into multiple `mkDerivation` arguments (not sure what the default should be but remember we want the `pkgsStatic`, `pkgsMusl` & `pkgsLLVM` variants to still build happily).

`version` and `src` are both overridable using `.overrideAttrs`, and can even be made to work "properly" if you use `finalAttrs` (see https://nixos.org/manual/nixpkgs/stable/#mkderivation-recursive-attributes) for any cross-references in the package rather than relying on `rec {` (which is evaluated at an early stage so doesn't work as well).

Reply:
Ah, I think I'm starting to grok what you're suggesting!
It helped a lot to see your PR #259070.
Thank you!

Removing `enableOddCrypto` completely is actually fine now that I reverted back to `--enable-sp-math-all`, unlike with `--enable-sp-math`.

I'll add a `variant`.

As HAProxy would need to both `.override` and `.overrideAttrs`, would calling them both be preferred to adding an `extraFlags ? []` or `extraOptions ? []`?

Reply:
In my opinion, yes - it moves quirkiness and complexity out of the main package and it doesn't _really_ do anything that can't be done with `.overrideAttrs`

Reply:
How would I remove flags like `--enable-writedup` or `--enable-pkcs11` without respecifying all the the other flags, having my overridden package fail every time extra options are added to the base derivation or having a (useless) derivation with none of the conflicting options?

Reply:
I wouldn't be too afraid to add specific file-level arguments for each, if you already have a use-case where you'll want to omit those flags.

In the worst case scenario without the dedicated arguments you can always `.overrideAttrs (prevAttrs: { configureFlags = lib.remove "--enable-writedup" prevAttrs.configureFlags; })` (a little ugly though)